In [1]:
import pandas as pd

In [2]:
def read_data(file_path):
    babies_data = []
    with open(file_path) as file:
        for line in file:
            babies_data.append(line.split())

    df = pd.DataFrame(babies_data[1:], columns = babies_data[0]).astype(int)

#     for col in df.columns:
#         df[col] = pd.to_numeric(df[col])
    return df

In [3]:
df = read_data('data/babies23.data')

In [12]:
df.describe().applymap(lambda x: f"{x:0.1f}")

id pluralty outcome    date gestation     sex      wt  parity  \
count  1236.0   1236.0  1236.0  1236.0    1236.0  1236.0  1236.0  1236.0   
mean   6000.8      5.0     1.0  1535.8     286.9     1.0   119.6     1.9   
std    2257.3      0.0     0.0   106.9      75.2     0.0    18.2     1.9   
min      15.0      5.0     1.0  1350.0     148.0     1.0    55.0     0.0   
25%    5286.2      5.0     1.0  1444.0     272.0     1.0   108.8     0.0   
50%    6729.5      5.0     1.0  1539.5     280.0     1.0   120.0     1.0   
75%    7583.2      5.0     1.0  1627.0     288.0     1.0   131.0     3.0   
max    9263.0      5.0     1.0  1714.0     999.0     1.0   176.0    13.0   

         race     age  ...   drace    dage     ded     dht     dwt marital  \
count  1236.0  1236.0  ...  1236.0  1236.0  1236.0  1236.0  1236.0  1236.0   
mean      3.2    27.4  ...     3.7    30.7     3.2    81.7   505.4     1.0   
std       4.2     6.5  ...     6.8     8.5     1.7    14.3   406.7     0.3   
min       0.0    15.0  ...     0.0    18.0     0.0    60.0   110.0     0.0   
25%       0.0    23.0  ...     0.0    25.0     2.0    70.0   165.0     1.0   
50%       3.0    26.0  ...     3.0    29.0     4.0    73.0   190.0     1.0   
75%       7.0    31.0  ...     7.0    35.0     5.0    99.0   999.0     1.0   
max      99.0    99.0  ...    99.0    99.0     9.0    99.0   999.0     5.0   

          inc   smoke    time  number  
count  1236.0  1236.0  1236.0  1236.0  
mean     13.2     0.9     1.7     2.6  
std      28.4     1.2     8.8     8.9  
min       0.0     0.0     0.0     0.0  
25%       2.0     0.0     0.0     0.0  
50%       4.0     1.0     1.0     1.0  
75%       7.0     1.0     1.0     3.0  
max      98.0     9.0    99.0    98.0  

[8 rows x 23 columns]

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 23 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   id         1236 non-null   int64
 1   pluralty   1236 non-null   int64
 2   outcome    1236 non-null   int64
 3   date       1236 non-null   int64
 4   gestation  1236 non-null   int64
 5   sex        1236 non-null   int64
 6   wt         1236 non-null   int64
 7   parity     1236 non-null   int64
 8   race       1236 non-null   int64
 9   age        1236 non-null   int64
 10  ed         1236 non-null   int64
 11  ht         1236 non-null   int64
 12  wt         1236 non-null   int64
 13  drace      1236 non-null   int64
 14  dage       1236 non-null   int64
 15  ded        1236 non-null   int64
 16  dht        1236 non-null   int64
 17  dwt        1236 non-null   int64
 18  marital    1236 non-null   int64
 19  inc        1236 non-null   int64
 20  smoke      1236 non-null   int64
 21  time       123

## Clean data

In [7]:
with open('data/babies.readme') as file:
        print(file.read())

Variables in data file
1. id - identification number
2. pluralty -  5= single fetus
3. outcome - 1= live birth that survived at least 28 days
4. date - birth date where 1096=January1,1961 
5. gestation - length of gestation in days
6. sex - infant's sex 1=male 2=female 9=unknown
7. wt -  birth weight in ounces (999 unknown)
8. parity - total number of previous pregnancies including
   fetal deaths and still births, 99=unknown
9. race - mother's race 0-5=white 6=mex 7=black 8=asian 9=mixed 99=unknown 
10 age - mother's age in years at termination of pregnancy, 99=unknown
11 ed - mother's education 0= less than 8th grade, 
   1 = 8th -12th grade - did not graduate, 
   2= HS graduate--no other schooling , 3= HS+trade,
  4=HS+some college 5= College graduate, 6&7 Trade school HS unclear, 9=unknown 
12 ht - mother's height in inches to the last completed inch
   99=unknown
13 wt - mother prepregnancy wt in pounds, 999=unknown 
14 drace - father's race, coding same as mother's race.
15 dage

### Todo - data cleaning
- remove unnecessary columns:
    - id, date
    
- replace 999 and 99 by nans
    - wt: replace 999 by nan
    - parity: replace 99 by nan
    - age: replace 99 by nan
    - ht: replace 99 by nan
    - dage: replace 99 by nan
    - dht: replace 99 by nan
    - dwt: replace 999 by nan
    - inc: replace 98 and 99 by nan
    
- replace categorical numbers by their string values
    - pluralty:  5 = single fetus
    - outcome: 1 = live birth that survived at least 28 days
    - sex: infant's sex 1=male 2=female 9=unknown
    - race: mother's race 0-5=white 6=mex 7=black 8=asian 9=mixed 99=unknown 
    - ed - mother's education 0= < 8th grade, 1 = 8th to 12th grade, 2= HS , 3= HS+trade, 4=HS+some college 5=     College graduate, 6&7 Trade school HS unclear, 9=unknown 
    - drace: fathers's race 0-5=white 6=mex 7=black 8=asian 9=mixed 99=unknown
    - ded - father's education 0= < 8th grade, 1 = 8th to 12th grade, 2= HS , 3= HS+trade, 4=HS+some college 5=     College graduate, 6&7 Trade school HS unclear, 9=unknown 
    - marital 1=married, 2= legally separated, 3= divorced, 4=widowed, 5=never married
    - smoke - does mother smoke? 0=never, 1= smokes now, 2=until current pregnancy, 3=once did, not now, 9=unknown
    - time - 0=never smoked, 1=still smokes, 2=during current preg, 3=within 1 yr, 4= 1 to 2 years ago,
    5= 2 to 3 yr ago, 6= 3 to 4 yrs ago, 7=5 to 9yrs ago, 8=10+yrs ago, 9=quit and don't know, 98=unknown, 99=not     asked 
    
- other data preprocessing
    - wt: convert weight from ounces to kg.
    - ht - convert inches in cm
    - wt: convert weight from pounds to kg.
    - dht - convert inches in cm
    - dwt: convert weight from pounds to kg.
    - number  - take the average  0=never, 1=1-4, 2=5-9, 3=10-14, 4=15-19, 5=20-29, 6=30-39, 7=40-60, 8=60+,     9=smoke but don't know, 98=unknown, 99=not asked

- change column names:
    - wt --> baby's weight
    - parity --> nb_prev_births
    - race --> mother's race
    - age --> mother's age
    - ed --> mother's education
    - ht --> mother's height
    - wt --> mother's weight
    - drace --> father's race
    - dage --> father's age
    - ded --> father's education
    - dht --> father's height
    - dwt --> father's weight